src/socketio_botmoves.ts

```diff
@@ -0,0 +1,154 @@
+    emit(socket, filterPlayerView(playerID, data));
+  };
+
+  const sendAll = (payload: any) => {
+    pubSub.publish(getPubSubChannelId(matchID), payload);
+  };
+
+  return { send, sendAll };
+};
+
+/** Copied from boardgame.io/dist/src/master/master.ts */
+export async function fetch(db: any, matchID: string, partial: Partial<{state: boolean, metadata: boolean}>) {
+  let state;
+  if (isSynchronous(db)) {
+    ({ state } = db.fetch(matchID, partial));
+  } else {
+    ({ state } = await db.fetch(matchID, partial));
+  }
+  return state;
+}
+
+/// Bot's playerID is '1', because the gameWrapper uses player '0' for the human player. 
+export const BOT_ID = '1';
+
+/** This is a special transport specifically designed for replacing a player's move
+ * with a bot's move.
+ * 
+ * BOT_ID represents the player to replace.
+ * 
+ * Modifying the server is also needed to fill the bot's slot in the lobby (see injectBots()).
+ */
+export class SocketIOButBotMoves extends SocketIO {
+  init(
+    app: any,
+    games: Game[],
+    origins: any
+  ) {
+    super.init(app, games, origins);
+
+    for (const game of games) {
+      const nsp = app._io.of(game.name);
+      const bot = new (botWrapper(strategy))({
+        enumerate: (game as any).ai.enumerate,
+        seed: game.seed,
+      });
+
+      /** This should be in sync with how socket data is communicated.
+       * See boardgame.io/dist/src/server/transport/socketio.ts
+       */
+      nsp.on('connection', (socket: any) => {
+        socket.on('update', async (...args: Parameters<any>) => {
+          // The arguments are stale: we react to a player's step
+          // But we are on the same API that reacts to it
+          // Basically we assume that a socket.on('update', ...)
+          // already updated the gamestate, making StateID and PlayerID stale
+          const [_, staleStateID, matchID, stalePlayerID] : any[] = args;
+          const matchQueue = this.getMatchQueue(matchID);
+          await matchQueue.add(async () => {
+            // These happen after the player stepped.
+            // The state is written to storage, and the server now returned
+            // the authorative state to the player.
+            // TODO do not load the result from storage, reuse from the redux?
+            // TODO try do not send an authorative state to the player...?
+            console.log("Bot moves");
+
+            if (stalePlayerID === BOT_ID) {
+              // Do not react to bot's turn
+              return;
+            }
+            const state = await fetch(app.context.db, matchID, {state: true});
+            if (state.ctx.phase !== 'play' || currentPlayer(state.ctx) !== BOT_ID) {
+              // Not a real action, possibly a failed move.
+              return;
+            }
+            const botAction = await bot.play(
+              state,
+              GetBotPlayer(state, {[BOT_ID]: bot}) as any
+            );
+            
+            const master = new Master(
```

Comment:
Maybe we can extract Master to `nsp.on`, but boardgame.io re-created this class every time

Reply:
Maybe open upstream issue?